radix-engine-interface/src/time/date_time.rs

```diff
@@ -0,0 +1,512 @@
+use crate::time::constants::*;
+use crate::time::Instant;
+use sbor::*;
+use sbor::rust::format;
+use sbor::rust::string::{String, ToString};
+
+const UNIX_EPOCH_YEAR: u32 = 1970;
+const SECONDS_IN_A_NON_LEAP_YEAR: u64 = 365 * 24 * 60 * 60;
+const SECONDS_IN_A_LEAP_YEAR: u64 = 366 * 24 * 60 * 60;
+const DAYS_PER_4Y: i64 = 365 * 4 + 1;
+const DAYS_PER_100Y: i64 = 365 * 100 + 24;
+const DAYS_PER_400Y: i64 = 365 * 400 + 97;
+const LEAP_YEAR_DAYS_IN_MONTHS: [u8; 12] = [31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
+// A shift (in seconds) from the Unix epoch (1970-01-01 00:00:00)
+// to a base date that is a multiple of a 400-year leap cycle.
+// Used in Instant -> DateTime conversion.
+// The date we're using is 2000-03-01 00:00:00, for two reasons:
+// a) It's a multiple of 400, to make it easier to work with leap years
+// b) We're also shifting the month to 1st March, so that
+//    the extra day on leap years is added to the last month (Feb),
+//    not in the middle of a year (makes some calculations easier)
+const SHIFT_FROM_UNIX_TIME_TO_MARCH_Y2K: i64 = 946684800 + 86400 * (31 + 29);
+const MIN_SUPPORTED_TIMESTAMP: i64 = -62135596800;
+const MAX_SUPPORTED_TIMESTAMP: i64 = 135536014634284799;
+
+#[derive(Encode, Decode, TypeId, PartialEq, Eq, Copy, Clone, Debug)]
+pub struct DateTime {
+    year: u32,
+    month: u8,
+    day_of_month: u8,
+    hour: u8,
+    minute: u8,
+    second: u8,
+}
+
+impl DateTime {
+    pub fn new(
+        year: u32,
+        month: u8,
+        day_of_month: u8,
+        hour: u8,
+        minute: u8,
+        second: u8,
+    ) -> Result<Self, String> {
```

Comment:
Worth adding a proper enum for all error codes and moving the user-friendly messages to `fmt::Display` impl?